README.txt
.. include:: doc/source/readme.txt

Documentation & Code Examples
=============================
The stompest API is `fully documented here <http://nikipore.github.com/stompest/>`_.


CHANGES.txt
* 1.0.4 - Bug fix thanks to [Njal Karevoll](https://github.com/nkvoll). No longer relies on newline after the null-byte frame separator. Library is now compatible with RabbitMQ stomp adapter.
* 1.1.1 - Support for binary messages.
* 1.1.2 - Fixed issue with stomper adding a space in ACK message-id header. ActiveMQ 5.6.0 no longer tolerates this.
* 2.0a1 - Complete redesign: feature-complete implementation of STOMP 1.0 and 1.1. Broker failover. Decoupled from [stomper](http://code.google.com/p/stomper/).
* 2.0a2 - Heart-beating.
* 2.0a3 - Implementation of STOMP 1.2 (except repeated header entries). Support for UTF-8 headers and escaping of header delimiting characters (STOMP 1.1 and 1.2).
* 2.0a4 - Improved STOMP protocol version. Minor incompatible changes of the low-level API (protocol.session and protocol.commands). Bugfixes headers encoding.
* 2.0a5 - Factored out stompest.twisted into a separate PyPI package to make the stompest library self-consistent.
* 2.1.0 - Generic Listener API.
* 2.1.1 - Minor bugfixes.
* 2.1.2 - Bugfix subscription handler.
* 2.1.3 - Bugfix thanks to [Peter McAlpine](https://github.com/rtkpmcalpine). More stable detection of localhost network addresses.
* 2.1.4 - Bugfix heart-beat negotiation.
* 2.1.5 - Improvement thanks to [Ray Ruvinskiy](https://github.com/rtkrruvinskiy). Avoid unintended inflation of messaging handling time.
* 2.1.6 - Bugfix thanks to [Cory Stegelmeier](https://github.com/rtkcstegel). Retry system calls when EINTR is received.
* 2.2.0 - Support for Python 3.3-3.5 thanks to [Petr Rašek](https://github.com/hydratk). Dropped support for Python 2.6.
* 2.2.1 - Minor bugfixes. API break: async client.disconnect() parameter changed from "failure" to "reason".
* 2.2.2 - <broken and deleted>
* 2.2.3 - Optimized STOMP parser adapted from David Wilson's [tinystompest](https://github.com/dw).
* 2.2.4 - Increased test coverage of STOMP parser and fixed the uncovered bugs.
* 2.2.5 - Increased test coverage of STOMP parser even more and fixed the uncovered bugs. Cleaned up the parser implementation.
* 2.2.6 - Increased test coverage of STOMP parser even more and fixed the uncovered bugs. Cleaned up everything and tuned it significantly.
* 2.3.0 - Added support for TLS/SSL thanks to [Caleb Hattingh](https://github.com/cjrh).


stompest/protocol/spec.py
import codecs
from stompest.error import StompProtocolError

class StompSpec(object):
    """This class hosts all constants related to the STOMP protocol specification in its various versions. There really isn't much to document, but you are invited to take a look at all available constants in the source code. Wait a minute ... one attribute is particularly noteworthy, name :attr:`DEFAULT_VERSION` --- which currently is :obj:`'1.0'` (but this may change in upcoming stompest releases, so you're advised to always explicitly define which STOMP protocol version you are going to use).
    
    .. seealso :: Specification of STOMP protocols `1.0 <http://stomp.github.com//stomp-specification-1.0.html>`_ and `1.1 <http://stomp.github.com//stomp-specification-1.1.html>`_, your favorite broker's documentation for additional STOMP headers.
    """
    # specification of the STOMP protocol: http://stomp.github.com//index.html
    VERSION_1_0, VERSION_1_1, VERSION_1_2 = '1.0', '1.1', '1.2'
    VERSIONS = [VERSION_1_0, VERSION_1_1, VERSION_1_2]
    DEFAULT_VERSION = VERSION_1_0

    ABORT = u'ABORT'
    ACK = u'ACK'
    BEGIN = u'BEGIN'
    COMMIT = u'COMMIT'
    CONNECT = u'CONNECT'
    DISCONNECT = u'DISCONNECT'
    NACK = u'NACK'
    SEND = u'SEND'
    STOMP = u'STOMP'
    SUBSCRIBE = u'SUBSCRIBE'
    UNSUBSCRIBE = u'UNSUBSCRIBE'

    CLIENT_COMMANDS = {
        VERSION_1_0: set([
            ABORT, ACK, BEGIN, COMMIT, CONNECT, DISCONNECT,
            SEND, SUBSCRIBE, UNSUBSCRIBE
        ]),
        VERSION_1_1: set([
            ABORT, ACK, BEGIN, COMMIT, CONNECT, DISCONNECT,
            NACK, SEND, STOMP, SUBSCRIBE, UNSUBSCRIBE
        ]),
        VERSION_1_2: set([
            ABORT, ACK, BEGIN, COMMIT, CONNECT, DISCONNECT,
            NACK, SEND, STOMP, SUBSCRIBE, UNSUBSCRIBE
        ])
    }

    CONNECTED = u'CONNECTED'
    ERROR = u'ERROR'
    MESSAGE = u'MESSAGE'
    RECEIPT = u'RECEIPT'

    SERVER_COMMANDS = {
        VERSION_1_0: set([CONNECTED, ERROR, MESSAGE, RECEIPT]),
        VERSION_1_1: set([CONNECTED, ERROR, MESSAGE, RECEIPT]),
        VERSION_1_2: set([CONNECTED, ERROR, MESSAGE, RECEIPT])
    }

    COMMANDS = dict(CLIENT_COMMANDS)
    for (version, commands) in SERVER_COMMANDS.iteritems():
        COMMANDS.setdefault(version, set()).update(commands)

    COMMANDS_BODY_ALLOWED = {
        VERSION_1_1: set([SEND, MESSAGE, ERROR]),
        VERSION_1_2: set([SEND, MESSAGE, ERROR])
    }

    CODECS = {  # for command and headers
        VERSION_1_0: 'ascii'
    }
    CODECS = dict([
        (version, codecs.lookup(CODECS.get(version, 'utf-8'))) for version in VERSIONS
    ])

    LINE_DELIMITER = '\n'
    STRIP_LINE_DELIMITER = {
        VERSION_1_2: '\r'
    }

    ESCAPE_CHARACTER = '\\'
    ESCAPED_CHARACTERS = {
        VERSION_1_0: {'\\': '\\', 'c': ':', 'n': '\n'},
        VERSION_1_1: {'\\': '\\', 'c': ':', 'n': '\n'},
        VERSION_1_2: {'\\': '\\', 'c': ':', 'n': '\n', 'r': '\r'}
    }
    COMMANDS_ESCAPE_EXCLUDED = {
        VERSION_1_0: COMMANDS[VERSION_1_0],
        VERSION_1_1: set([CONNECT, CONNECTED]),
        VERSION_1_2: set([CONNECT, CONNECTED])
    }

    FRAME_DELIMITER = '\x00'
    HEADER_SEPARATOR = ':'

    ACCEPT_VERSION_HEADER = u'accept-version'
    ACK_HEADER = u'ack'
    CONTENT_LENGTH_HEADER = u'content-length'
    CONTENT_TYPE_HEADER = u'content-type'
    DESTINATION_HEADER = u'destination'
    HEART_BEAT_HEADER = u'heart-beat'
    HOST_HEADER = u'host'
    ID_HEADER = u'id'
    LOGIN_HEADER = u'login'
    MESSAGE_ID_HEADER = u'message-id'
    PASSCODE_HEADER = u'passcode'
    RECEIPT_HEADER = u'receipt'
    RECEIPT_ID_HEADER = u'receipt-id'
    SESSION_HEADER = u'session'
    SERVER_HEADER = u'server'
    SUBSCRIPTION_HEADER = u'subscription'
    TRANSACTION_HEADER = u'transaction'
    VERSION_HEADER = u'version'

    ACK_AUTO = u'auto'
    ACK_CLIENT = u'client'
    ACK_CLIENT_INDIVIDUAL = u'client-individual'
    CLIENT_ACK_MODES = set([ACK_CLIENT, ACK_CLIENT_INDIVIDUAL])

    HEART_BEAT_SEPARATOR = ','

    @classmethod
    def version(cls, version=None):
        """Check whether **version** is a valid STOMP protocol version.
        
        :param version: A candidate version, or :obj:`None` (which is equivalent to the value of :attr:`StompSpec.DEFAULT_VERSION`). 
        """
        if version is None:
            version = cls.DEFAULT_VERSION
        if version not in cls.VERSIONS:
            raise StompProtocolError('Version is not supported [%s]' % version)
        return version

    @classmethod
    def versions(cls, version):
        """Obtain all versions prior or equal to **version**.
        """
        version = cls.version(version)
        for v in cls.VERSIONS:
            yield v
            if v == version:
                break

